### PR TITLE
Plane passing through three points

### PR DESCRIPTION
**Pull request by stla**
Needed for the torus passing through three points.

I also implemented the `in` function for checking whether a point belongs to a plane, because it is the appropriate thing to do for the unit tests, and it is interesting in itself anyway.

### CODE REVIEW

File: src/primitives/plane.jl [hunk 1]
```diff
@@ -30,6 +30,15 @@ Plane(p::Tuple, u::Tuple, v::Tuple) = Plane(Point(p), Vec(u), Vec(v))
 
 Plane(p::Tuple, n::Tuple) = Plane(Point(p), Vec(n))
 
+function Plane(p1::Point{3,T1}, p2::Point{3,T2}, p3::Point{3,T3}) where {T1,T2,T3}
+  T = promote_type(T1, T2, T3)
+  n⃗ = Vec{3,T}((p2 - p1) × (p3 - p1))
+  if isapprox(norm(n⃗), zero(T); atol = atol(T))
```
Comment (juliohm) [inline src/primitives/plane.jl]:
```suggestion
  T = promote_type(T1, T2, T3)
  if isapprox(area(Triangle(p1, p2, p3)), zero(T), atol = atol(T))
```

Comment (stla) [reply src/primitives/plane.jl]: Well seen!

File: src/primitives/plane.jl [hunk 2]
```diff
@@ -60,4 +69,11 @@ normal(p::Plane) = normalize(p.u × p.v)
...
 ==(p₁::Plane{T}, p₂::Plane{T}) where {T} =
   isapprox((p₁.v - p₁.u) ⋅ normal(p₂), zero(T), atol=atol(T)) &&
-  isapprox((p₂.v - p₂.u) ⋅ normal(p₁), zero(T), atol=atol(T))
\ No newline at end of file
+  isapprox((p₂.v - p₂.u) ⋅ normal(p₁), zero(T), atol=atol(T))
+
+function Base.in(pt::Point, pl::Plane)
+  dotprod = normal(pl) ⋅ (pt - origin(pl))
+  T = typeof(dotprod)
+  isapprox(dotprod, zero(T); atol=atol(T))
```
Comment (juliohm) [inline src/primitives/plane.jl]:
```suggestion
  d = normal(pl) ⋅ (pt - origin(pl))
  T = typeof(d)
  isapprox(d, zero(T), atol = atol(T))
```

**Review by juliohm — CHANGES_REQUESTED** [review]
Review attached.

File: src/primitives/plane.jl (hunk 2 above)
Comment (stla) [reply src/primitives/plane.jl]: A comma to separate `atol`? Isn't it a keyword argument? In the help they always use a semi-colon.

Comment (juliohm) [reply src/primitives/plane.jl]: It is a matter of taste I guess. Only in specific cases the semi-colon makes a difference. So if we can stick to simple commas by default that would be more consistent with the rest of the codebase.

File: src/primitives/plane.jl [hunk 3]
```diff
@@ -30,6 +30,15 @@ Plane(p::Tuple, u::Tuple, v::Tuple) = Plane(Point(p), Vec(u), Vec(v))
 
 Plane(p::Tuple, n::Tuple) = Plane(Point(p), Vec(n))
 
+function Plane(p1::Point{3,T1}, p2::Point{3,T2}, p3::Point{3,T3}) where {T1,T2,T3}
+  T = promote_type(T1, T2, T3)
+  n⃗ = Vec{3,T}((p2 - p1) × (p3 - p1))
+  if isapprox(area(Triangle(p1, p2, p3)), zero(T), atol = atol(T))
```
Comment (juliohm) [inline src/primitives/plane.jl]:
```suggestion
  t = Triangle(p1, p2, p3)
  n = normal(t)
  if isapprox(area(t), zero(T), atol = atol(T))
```

File: src/primitives/plane.jl [hunk 4]
```diff
@@ -30,6 +30,15 @@ Plane(p::Tuple, u::Tuple, v::Tuple) = Plane(Point(p), Vec(u), Vec(v))
...
+    throw(ArgumentError("The three points are colinear."))
+  end
+  Plane(convert(Point{3,T}, p1), n⃗)
```
Comment (juliohm) [inline src/primitives/plane.jl]:
```suggestion
  Plane(p1, n)
```

File: src/primitives/plane.jl [hunk 5]
```diff
@@ -30,6 +30,15 @@ Plane(p::Tuple, u::Tuple, v::Tuple) = Plane(Point(p), Vec(u), Vec(v))
 
 Plane(p::Tuple, n::Tuple) = Plane(Point(p), Vec(n))
 
+function Plane(p1::Point{3,T1}, p2::Point{3,T2}, p3::Point{3,T3}) where {T1,T2,T3}
+  T = promote_type(T1, T2, T3)
```
Comment (juliohm) [inline src/primitives/plane.jl]:
```suggestion
function Plane(p1::Point{3,T}, p2::Point{3,T}, p3::Point{3,T}) where {T}
```

File: src/primitives/plane.jl [hunk 6]
```diff
@@ -30,6 +30,15 @@ Plane(p::Tuple, u::Tuple, v::Tuple) = Plane(Point(p), Vec(u), Vec(v))
 
 Plane(p::Tuple, n::Tuple) = Plane(Point(p), Vec(n))
 
+function Plane(p1::Point{3,T}, p2::Point{3,T}, p3::Point{3,T}) where {T}
+  t = Triangle(p1, p2, p3)
+  n = normal(t)
+  if isapprox(area(t), zero(T), atol = atol(T))
+    throw(ArgumentError("The three points are colinear."))
+  end
+  Plane(p1, n)
```
Comment (juliohm) [inline src/primitives/plane.jl]:
```suggestion
  if isapprox(area(t), zero(T), atol = atol(T))
    throw(ArgumentError("The three points are colinear."))
  end
  Plane(p1, normal(t))
```

File: src/primitives/plane.jl [hunk 7]
```diff
@@ -60,4 +68,11 @@ normal(p::Plane) = normalize(p.u × p.v)
 
 ==(p₁::Plane{T}, p₂::Plane{T}) where {T} =
   isapprox((p₁.v - p₁.u) ⋅ normal(p₂), zero(T), atol=atol(T)) &&
-  isapprox((p₂.v - p₂.u) ⋅ normal(p₁), zero(T), atol=atol(T))
\ No newline at end of file
+  isapprox((p₂.v - p₂.u) ⋅ normal(p₁), zero(T), atol=atol(T))
+
+function Base.in(pt::Point, pl::Plane)
+  d = normal(pl) ⋅ (pt - origin(pl))
+  T = typeof(d)
```
Comment (juliohm) [inline src/primitives/plane.jl]:
```suggestion
function Base.in(pt::Point{3,T}, pl::Plane{T}) where {T}
  d = normal(pl) ⋅ (pt - origin(pl))
```

File: src/primitives/plane.jl [hunk 8]
```diff
@@ -60,4 +68,10 @@ normal(p::Plane) = normalize(p.u × p.v)
...
 ==(p₁::Plane{T}, p₂::Plane{T}) where {T} =
   isapprox((p₁.v - p₁.u) ⋅ normal(p₂), zero(T), atol=atol(T)) &&
-  isapprox((p₂.v - p₂.u) ⋅ normal(p₁), zero(T), atol=atol(T))
\ No newline at end of file
+  isapprox((p₂.v - p₂.u) ⋅ normal(p₁), zero(T), atol=atol(T))
+
+function Base.in(pt::Point{3,T}, pl::Plane{T}) where {T}
+  d = normal(pl) ⋅ (pt - origin(pl))
+  isapprox(d, zero(T), atol = atol(T))
+end
```
Comment (juliohm) [inline src/primitives/plane.jl]:
```suggestion
Base.in(pt::Point{3,T}, pl::Plane{T}) where {T} =
  isapprox(normal(pl) ⋅ (pt - origin(pl)), zero(T), atol = atol(T))
```

**Review by juliohm — APPROVED** [review]

File: src/primitives/plane.jl [hunk 9]
```diff
@@ -60,4 +68,8 @@ normal(p::Plane) = normalize(p.u × p.v)
 
 ==(p₁::Plane{T}, p₂::Plane{T}) where {T} =
   isapprox((p₁.v - p₁.u) ⋅ normal(p₂), zero(T), atol=atol(T)) &&
-  isapprox((p₂.v - p₂.u) ⋅ normal(p₁), zero(T), atol=atol(T))
\ No newline at end of file
+  isapprox((p₂.v - p₂.u) ⋅ normal(p₁), zero(T), atol=atol(T))
+
+Base.in(pt::Point{3,T}, pl::Plane{T}) where {T} =
+  isapprox(normal(pl) ⋅ (pt - origin(pl)), zero(T), atol = atol(T))
+
```
Comment (juliohm) [inline src/primitives/plane.jl]:
```suggestion
```

**Pull request merged**